KomaMRIBase/src/datatypes/phantom/motion/ArbitraryMotion.jl

```diff
@@ -43,106 +60,87 @@ julia> motion = ArbitraryMotion(
+struct ArbitraryMotion{T} <: MotionModel{T}
+    t_start::T
+    t_end::T
+    dx::AbstractArray{T}
+    dy::AbstractArray{T}
+    dz::AbstractArray{T}
 end
 
 function Base.getindex(
     motion::ArbitraryMotion, p::Union{AbstractRange,AbstractVector,Colon}
 )
-    fields = []
-    for field in fieldnames(ArbitraryMotion)
-        if field in (:dx, :dy, :dz)
-            push!(fields, getfield(motion, field)[p, :])
-        elseif field in (:ux, :uy, :uz)
-            push!(fields, getfield(motion, field)[p])
-        else
-            push!(fields, getfield(motion, field))
-        end
-    end
-    return ArbitraryMotion(fields...)
+    return ArbitraryMotion(motion.t_start, motion.t_end, motion.dx[p,:], motion.dy[p,:], motion.dz[p,:])
 end
 
 Base.:(==)(m1::ArbitraryMotion, m2::ArbitraryMotion) = reduce(&, [getfield(m1, field) == getfield(m2, field) for field in fieldnames(ArbitraryMotion)])
 Base.:(≈)(m1::ArbitraryMotion, m2::ArbitraryMotion)  = reduce(&, [getfield(m1, field) ≈ getfield(m2, field) for field in fieldnames(ArbitraryMotion)])
 
 function Base.vcat(m1::ArbitraryMotion, m2::ArbitraryMotion)
     fields = []
-    @assert m1.period_durations == m2.period_durations "period_durations of both ArbitraryMotions must be the same"
-    for field in
-        Iterators.filter(x -> !(x == :period_durations), fieldnames(ArbitraryMotion))
+    @assert (m1.t_start == m2.t_start) && (m1.t_end == m2.t_end) "starting and ending times must be the same"
+    for field in (:dx, :dy, :dz)
         push!(fields, [getfield(m1, field); getfield(m2, field)])
     end
-    return ArbitraryMotion(m1.period_durations, fields...)
+    return ArbitraryMotion(m1.t_start, m1.t_end, fields...)
 end
 
 """
     limits = times(obj.motion)
 """
 function times(motion::ArbitraryMotion)
-    period_durations = motion.period_durations
-    num_pieces = size(motion.dx)[2] + 1
-    return times(period_durations, num_pieces)
+    return collect(range(motion.t_start, motion.t_end, length=size(motion.dx)[2]))
 end
 
-function times(period_durations::AbstractVector, num_pieces::Int)
-    # Pre-allocating memory
-    limits = zeros(eltype(period_durations), num_pieces * length(period_durations) + 1)
-
-    idx = 1
-    for i in 1:length(period_durations)
-        segment_increment = period_durations[i] / num_pieces
-        cumulative_sum = limits[idx]  # Start from the last computed value in limits
-        for j in 1:num_pieces
-            cumulative_sum += segment_increment
-            limits[idx + 1] = cumulative_sum
-            idx += 1
-        end
+
+function get_itp_functions(
+    dx::AbstractArray{T},
+    dy::AbstractArray{T},
+    dz::AbstractArray{T}
+    ) where {T<:Real}
+    Ns, Nt = size(dx)
+
+    t = similar(dx, Nt)
+    t .= 0:(Nt-1)
+
+    id = similar(dx, Ns)
+    id .= 1:Ns
+
+    itpx = GriddedInterpolation((id, t ./ (Nt-1)), dx)
+    itpy = GriddedInterpolation((id, t ./ (Nt-1)), dy)
+    itpz = GriddedInterpolation((id, t ./ (Nt-1)), dz)
```

Comment:
> `t .= 0:(Nt-1)` -> `t .= range(0,1,Nt)`?

Solved in [5050c46](https://github.com/JuliaHealth/KomaMRI.jl/pull/408/commits/5050c46c13a69ef0049e255837b0e3eddf346b93) and [9ab9abb](https://github.com/JuliaHealth/KomaMRI.jl/pull/408/commits/9ab9abbe6343905dfd846793199570fd3310f7ce)

> Thinking about it, I don't get why `t` and `id` need to be GPU arrays; if they were UnitRanges with eltypes `Int` and `T`, the generators would use almost 0 memory. Wouldn't this be faster?

Solved  in [b8a2b83](https://github.com/JuliaHealth/KomaMRI.jl/pull/408/commits/b8a2b83a040dcffe8b239c347cb4f58689c092fc). It is neither faster nor slower

Reply:
Then that means that we can use `linear_interpolation`, and not define any custom interpolation function.

Reply:
Using `interpolation`:

```julia
function interpolate(motion::ArbitraryMotion{T}) where {T<:Real}
    Ns, Nt = size(motion.dx)
    itpx = Interpolations.interpolate((1:Ns, range(zero(T), oneunit(T), Nt)), motion.dx, (NoInterp(), Gridded(Linear())))
    itpy = Interpolations.interpolate((1:Ns, range(zero(T), oneunit(T), Nt)), motion.dy, (NoInterp(), Gridded(Linear())))
    itpz = Interpolations.interpolate((1:Ns, range(zero(T), oneunit(T), Nt)), motion.dz, (NoInterp(), Gridded(Linear())))
    return itpx, itpy, itpz
end
```

```julia
function resample(
    itpx::Interpolator{T}, 
    itpy::Interpolator{T}, 
    itpz::Interpolator{T}, 
    t::AbstractArray{T}
) where {T<:Real}
    Ns =size(itpx.coefs,1)
    return itpx.(1:Ns, t), itpy.(1:Ns, t), itpz.(1:Ns, t)
end
```

throws scalar indexing error when calling itpx.(1:Ns, t), so the custom function is necessary.

Reply:
```julia-repl
julia> using CUDA, Interpolations

julia> Ns = 1_000; Nt = 100;

julia> itp = linear_interpolation((1:Ns, range(0f0, 1f0, Nt)), cu(randn(Ns, Nt)));

julia> itp.(1:Ns, (range(0.3f0, 0.5f0, 10))');

julia> itp.(1:Ns, cu(range(0.3f0, 0.5f0, 10))');
```

I don't get scalar indexing errors.

Reply:
Sorry this one is better

```julia
julia> itp = linear_interpolation((1:Ns, range(0f0, 1f0, Nt)), randn(Ns, Nt));

julia> itp_gpu = cu(itp);

julia> tq = collect(range(0.3, 0.5, 10));

julia> res_gpu = itp_gpu.(1:Ns, cu(tq)');

julia> res_cpu = itp.(1:Ns, tq');

julia> res_cpu .- Array(res_gpu)
1000×10 Matrix{Float64}:
 1.1582e-6  -2.62224e-6   5.36099e-6  6.77039e-7  …  -1.57927e-6  -5.04696e-7   4.7508e-8
 ⋮                                                ⋱
 1.2986e-6  -2.33043e-6  -5.2257e-6   3.63712e-8     -1.76235e-7  -1.41588e-7  -2.1834e-8
```

Reply:
With Ns = 100_000

```julia-repl

julia> @CUDA.time itp.(1:Ns, tq');
  0.007852 seconds (5 CPU allocations: 7.630 MiB)

julia> @CUDA.time itp_gpu.(1:Ns, cu(tq)');
  0.000282 seconds (91 CPU allocations: 2.344 KiB) (2 GPU allocations: 7.629 MiB, 3.97% memmgmt time)
```

Reply:
> ```julia-repl
> julia> using CUDA, Interpolations
> 
> julia> Ns = 1_000; Nt = 100;
> 
> julia> itp = linear_interpolation((1:Ns, range(0f0, 1f0, Nt)), cu(randn(Ns, Nt)));
> 
> julia> itp.(1:Ns, (range(0.3f0, 0.5f0, 10))');
> 
> julia> itp.(1:Ns, cu(range(0.3f0, 0.5f0, 10))');
> ```
> 
> I don't get scalar indexing errors.

Both of them return a CPU Array, not in GPU. 

> Sorry this one is better (Ns = 1000)
> 
> ```julia
> julia> itp = linear_interpolation((1:Ns, range(0f0, 1f0, Nt)), randn(Ns, Nt));
> 
> julia> itp_gpu = cu(itp);
> 
> julia> tq = collect(range(0.3, 0.5, 10));
> 
> julia> res_gpu = itp_gpu.(1:Ns, cu(tq)');
> 
> julia> res_cpu = itp.(1:Ns, tq');
> 
> julia> res_cpu .- Array(res_gpu)
> 1000×10 Matrix{Float64}:
>  1.1582e-6  -2.62224e-6   5.36099e-6  6.77039e-7  …  -1.57927e-6  -5.04696e-7   4.7508e-8
>  ⋮                                                ⋱
>  1.2986e-6  -2.33043e-6  -5.2257e-6   3.63712e-8     -1.76235e-7  -1.41588e-7  -2.1834e-8
> 
> julia> @CUDA.time itp.(1:Ns, tq');
>   0.000095 seconds (5 CPU allocations: 78.328 KiB)
> 
> julia> @CUDA.time itp_gpu.(1:Ns, cu(tq)');
>   0.000119 seconds (89 CPU allocations: 2.312 KiB) (2 GPU allocations: 78.164 KiB, 10.47% memmgmt time)
> ```

Here you first create the interpolating function from CPU arrays and then pass it to GPU. Should I go back to that?

Reply:
I know, they are both in the cpu because they are ranges (generators), that is what I'm trying to say only the chiefs are in the GPU, so no scalar indexing problems. You always had arrays (allocations). I'm no sure which combination is faster, you need to test it.

Reply:
Solved in [872d16a](https://github.com/JuliaHealth/KomaMRI.jl/pull/408/commits/872d16a09b844b0bc0a62f02cf0bdc61ffd5f2b3).
The only problem is the range `id`, it has to be `one(T):Ns` (and not `1:Ns`) for now, since:

```julia
using KomaMRI, Interpolations, CUDA
dx = gpu(randn(2,10), CUDABackend())
Ns, Nt = size(dx)
id = 1:Ns
t  = range(zero(eltype(dx)), oneunit(eltype(dx)), Nt)
itpx = KomaMRIBase.GriddedInterpolation((id, t), motion.dx, (NoInterp(), Gridded(Linear())));

tp = gpu(t, CUDABackend())
itpx.(id, tp')
```

Gives the following error:
```
ERROR: InvalidIRError: compiling MethodInstance for (::GPUArrays.var"#35#37")(::CUDA.CuKernelContext, ::CuDeviceMatrix{…}, ::Base.Broadcast.Broadcasted{…}, ::Int64) resulted in invalid LLVM IR
Reason: unsupported call to an unknown function (call to julia.new_gc_frame)
Reason: unsupported call to an unknown function (call to julia.push_gc_frame)
Reason: unsupported call to an unknown function (call to julia.pop_gc_frame)
Reason: unsupported call to an unknown function (call to julia.get_gc_frame_slot)
Reason: unsupported dynamic function invocation (call to _broadcast_getindex_evalf(f::Tf, args::Vararg{Any, N}) where {Tf, N} @ Base.Broadcast broadcast.jl:709)
Stacktrace:
 [1] _broadcast_getindex
   @ ./broadcast.jl:682
 [2] getindex
   @ ./broadcast.jl:636
 [3] #35
   @ ~/.julia/packages/GPUArrays/HjWFN/src/host/broadcast.jl:70
Hint: catch this exception as `err` and call `code_typed(err; interactive = true)` to introspect the erronous code with Cthulhu.jl
Stacktrace:
  [1] check_ir(job::GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget, CUDA.CUDACompilerParams}, args::LLVM.Module)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/kqxyC/src/validation.jl:147
  [2] macro expansion
    @ ~/.julia/packages/GPUCompiler/kqxyC/src/driver.jl:445 [inlined]
  [3] macro expansion
    @ ~/.julia/packages/TimerOutputs/Lw5SP/src/TimerOutput.jl:253 [inlined]
  [4] macro expansion
    @ ~/.julia/packages/GPUCompiler/kqxyC/src/driver.jl:444 [inlined]
  [5] 
    @ GPUCompiler ~/.julia/packages/GPUCompiler/kqxyC/src/utils.jl:92
  [6] emit_llvm
    @ ~/.julia/packages/GPUCompiler/kqxyC/src/utils.jl:86 [inlined]
  [7] 
    @ GPUCompiler ~/.julia/packages/GPUCompiler/kqxyC/src/driver.jl:134
  [8] codegen
    @ ~/.julia/packages/GPUCompiler/kqxyC/src/driver.jl:115 [inlined]
  [9] 
    @ GPUCompiler ~/.julia/packages/GPUCompiler/kqxyC/src/driver.jl:111
 [10] compile
    @ ~/.julia/packages/GPUCompiler/kqxyC/src/driver.jl:103 [inlined]
 [11] #1145
    @ ~/.julia/packages/CUDA/75aiI/src/compiler/compilation.jl:254 [inlined]
 [12] JuliaContext(f::CUDA.var"#1145#1148"{GPUCompiler.CompilerJob{…}}; kwargs::@Kwargs{})
    @ GPUCompiler ~/.julia/packages/GPUCompiler/kqxyC/src/driver.jl:52
 [13] JuliaContext(f::Function)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/kqxyC/src/driver.jl:42
 [14] compile(job::GPUCompiler.CompilerJob)
    @ CUDA ~/.julia/packages/CUDA/75aiI/src/compiler/compilation.jl:253
 [15] actual_compilation(cache::Dict{…}, src::Core.MethodInstance, world::UInt64, cfg::GPUCompiler.CompilerConfig{…}, compiler::typeof(CUDA.compile), linker::typeof(CUDA.link))
    @ GPUCompiler ~/.julia/packages/GPUCompiler/kqxyC/src/execution.jl:128
 [16] cached_compilation(cache::Dict{…}, src::Core.MethodInstance, cfg::GPUCompiler.CompilerConfig{…}, compiler::Function, linker::Function)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/kqxyC/src/execution.jl:103
 [17] macro expansion
    @ ~/.julia/packages/CUDA/75aiI/src/compiler/execution.jl:369 [inlined]
 [18] macro expansion
    @ ./lock.jl:267 [inlined]
 [19] cufunction(f::GPUArrays.var"#35#37", tt::Type{Tuple{…}}; kwargs::@Kwargs{})
    @ CUDA ~/.julia/packages/CUDA/75aiI/src/compiler/execution.jl:364
 [20] cufunction
    @ ~/.julia/packages/CUDA/75aiI/src/compiler/execution.jl:361 [inlined]
 [21] macro expansion
    @ ~/.julia/packages/CUDA/75aiI/src/compiler/execution.jl:112 [inlined]
 [22] #launch_heuristic#1204
    @ ~/.julia/packages/CUDA/75aiI/src/gpuarrays.jl:17 [inlined]
 [23] launch_heuristic
    @ ~/.julia/packages/CUDA/75aiI/src/gpuarrays.jl:15 [inlined]
 [24] _copyto!
    @ ~/.julia/packages/GPUArrays/HjWFN/src/host/broadcast.jl:78 [inlined]
 [25] copyto!
    @ ~/.julia/packages/GPUArrays/HjWFN/src/host/broadcast.jl:44 [inlined]
 [26] copy
    @ ~/.julia/packages/GPUArrays/HjWFN/src/host/broadcast.jl:29 [inlined]
 [27] materialize(bc::Base.Broadcast.Broadcasted{…})
    @ Base.Broadcast ./broadcast.jl:903
 [28] top-level scope
    @ ~/koma_tests/test_arbitrary.jl:267
Some type information was truncated. Use `show(err)` to see complete types.
```